Managed Identity in MSALJS

Robbie-Microsoft opened the pull request:
Implemented Managed Identity in MSALJS. The following sources can be used to create a Managed Identity application:
- App Service
- Azure Arc
- Cloud Shell
- Imds
- Service Fabric

Tested all sources manually, and have unit tests which make fake network requests.

Robbie-Microsoft committed:
MSAL Node's Managed Identity Api Implementation (#6514)

This PR covers the API for the new (MSAL-Node specific) Managed Identity
Client. Only the App Service source has been implemented for now. The
Azure Arc, CloudShell, IMDS and Service Fabric sources will be in
another PR.

The following is how a developer will create a new managed identity
client and acquire a token

```
import * as msal from "@azure/msal-node";

// optional
const config: msal.ManagedIdentityConfiguration = {
    // optional
    managedIdentityIdParams: {
        userAssignedClientId: "managed_identity_id"
    },
    system: { // optional
        loggerOptions: msal.LoggerOptions; // optional
        networkClient: msal.INetworkModule; // optional
        proxyUrl: string; // optional
        customAgentOptions: http.AgentOptions | https.AgentOption; // optional
    },
};

// required
const request: msal.ManagedIdentityRequest = {
    forceRefresh: true, // optional, defaults to false
    resource: "URI_of_the_managed_identity_assigned_to_the_azure_resource" // required
};

..................

try {
    const managedIdentityClient = new msal.ManagedIdentityClient(config);
    const response = await managedIdentityClient.acquireToken(request);
    console.log(response);
} catch (error) {
    console.log(error);
}
```

---------

Co-authored-by: Robbie Ginsburg <rgins16@gmail.com>

Robbie-Microsoft committed:
Implementation of Azure Arc Managed Identity into MSAL-Node's MSI Feature Branch (#6681)

Co-authored-by: Robbie Ginsburg <rgins16@gmail.com>

Robbie-Microsoft committed:
Separated Managed Identity Unit Tests into Multiple Files, Depending on the Source Type (#6751)

Robbie-Microsoft committed:
Implemented Static Caching + Code Cleanup (#6755)

Robbie-Microsoft committed:
Implemented GitHub feedback that was supposed to be included in the static caching commit

Robbie-Microsoft committed:
Managed Identity Documentation (#6763)

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Fixed bugs and linter errors caused from merge conflicts

Robbie-Microsoft committed:
Managed Identity Azure VM Sample (#6767)

Successfully ran sample on an Azure VM

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Updated reference to alpha package

Robbie-Microsoft committed:
Updated reference to alpha package

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Implemented Service Fabric + Cloud Shell Managed Identity (#6780)

In making the Service Fabric + Cloud Shell code as efficient as
possible, several code improvements that affect the other managed
identity sources were implemented as well.

Cloud Shell has been manually tested
Service Fabric still needs to be manually tested

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

bgavrilMS commented on lib/msal-node/src/network/HttpClient.ts:
```diff
@@ -20,13 +20,16 @@ import https from "https";
 export class HttpClient implements INetworkModule {
     private proxyUrl: string;
     private customAgentOptions: http.AgentOptions | https.AgentOptions;
+    private managedIdentity: boolean;
```
HttpClient should NOT know anything about Managed Identity. This is a violation of the Single Responsability Principle. 

At most, HTTPClient should know that it has to use the "http" module vs the "https" module. 

Can't you just look at the request URL to tell if the request URL starts with "http" and not "https" ?

bgavrilMS commented on lib/msal-node/src/config/Configuration.ts:
```diff
@@ -199,3 +214,34 @@ export function buildAppConfiguration({
+            system?.customAgentOptions as
+                | http.AgentOptions
+                | https.AgentOptions,
+            true // Managed Identity
```
Please don't leak MSI details in network client.

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

bgavrilMS commented on lib/msal-node/src/client/ManagedIdentityClient.ts:
```diff
@@ -0,0 +1,128 @@
+    private networkClient: INetworkModule;
+    private cryptoProvider: CryptoProvider;
+
+    static identitySource:
```
Pls add visibility modifier (private?) all the time.

bgavrilMS commented on lib/msal-node/src/client/ManagedIdentityClient.ts:
```diff
@@ -0,0 +1,128 @@
+        | CloudShell
+        | AzureArc
+        | Imds
+        | undefined;
```
Please stop using "undefined" and use null checks. https://stackoverflow.com/questions/5076944/what-is-the-difference-between-null-and-undefined-in-javascript

An identity source exists, but the variable can be null at first, before it is initialized.

Robbie-Microsoft replied:
I will stop using it in this case, but there are other cases in this PR where the variables can be undefined.

bgavrilMS commented on lib/msal-node/src/client/ManagedIdentityClient.ts:
```diff
@@ -0,0 +1,128 @@
+    private cryptoProvider: CryptoProvider;
+
+    static identitySource:
+        | ServiceFabric
```
Don't enumerate the classes here. Use `BaseManagedIdentitySource` instead - the base class. That's the power of polymorphism.

Robbie-Microsoft committed:
Implemented HttpClient-Decorator Retry-Policy in MSAL-Node & Managed Identity (#6891)

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Change files

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Updated package-lock

Robbie-Microsoft committed:
Improved unit test: proactively refreshes a token in the background when its refresh_in value is expired.

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Updated MSI sample to use access token to get a secret from a key vault (#6944)

Tested on Azure VM.

Will post these updates to the same sample
[here](https://github.com/Azure-Samples/msal-managed-identity/tree/main/src/nodejs)
after this gets approved.

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Changed the Managed Identity retry policy's number of times to retry from 1 to 3 (#6950)

Updated unit tests

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Update MSI Sample (#6955)

Tested on Azure VM

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Managed Identity in MSALJS - Final Improvements (#6965)

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Added Service Fabric Sample (#6966)

This sample is almost identical to the Azure VM sample. Only the README
really needs to be reviewed.

Robbie-Microsoft committed:
Updated package-lock.json

Robbie-Microsoft committed:
Updated README

Robbie-Microsoft committed:
Updated package-lock.json

pmaytak reviewed (approved):
Approving in lieu of Neha's approval.

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Updated package-lock.json

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Updated expressjs package

Robbie-Microsoft committed:
Updated package-lock.json

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Updated package-lock.json

Robbie-Microsoft committed:
bumped @azure/identity to 3.4.2

Robbie-Microsoft committed:
deleted samples

Robbie-Microsoft committed:
added samples back + updated package-lock.json

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
updated package-lock.json

Robbie-Microsoft committed:
Merge branch 'msi_feature_branch' of https://github.com/AzureAD/microsoft-authentication-library-for-js into msi_feature_branch

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
Pinned identity version in e2e samples

Robbie-Microsoft committed:
Updated package-lock.json

Robbie-Microsoft committed:
Merge branch 'dev' into msi_feature_branch

Robbie-Microsoft committed:
updatde package-lock.json
